src/main/java/io/jenkins/plugins/httpclient/RobustHTTPClient.java

```diff
@@ -0,0 +1,281 @@
+                                    } else {
+                                        diag = null;
+                                    }
+                                    throw new AbortException(String.format("Failed to %s, response: %d %s, body: %s", whatVerbose, responseCode.get(), statusLine != null ? statusLine.getReasonPhrase() : "?", diag));
+                                }
+                                connectionUser.use(response);
+                            }
+                        }
+                        return null; // success
+                    }).get(timeout, TimeUnit.SECONDS);
+                } catch (TimeoutException x) {
+                    throw new ExecutionException(new IOException(x)); // ExecutionException unwrapped & treated as retryable below
+                }
+                listener.getLogger().flush(); // seems we can get interleaved output with master otherwise
+                return; // success
+            } catch (ExecutionException wrapped) {
+                Throwable x = wrapped.getCause();
+                if (x instanceof IOException) {
+                    if (attempt == stopAfterAttemptNumber) {
+                        throw (IOException) x; // last chance
+                    }
+                    if (responseCode.get() > 0 && responseCode.get() < 200 || responseCode.get() >= 300 && responseCode.get() < 500) {
+                        throw (IOException) x; // 4xx errors should not be retried
+                    }
+                    // TODO exponent base could be made into a configurable parameter
+                    Thread.sleep(Math.min(((long) Math.pow(2d, attempt)) * waitMultiplier, waitMaximum * 1000));
+                    listener.getLogger().printf("Retrying %s after: %s%n", whatConcise, x instanceof AbortException ? x.getMessage() : x.toString());
+                    attempt++; // and continue
+                } else if (x instanceof InterruptedException) { // all other exceptions considered fatal
+                    throw (InterruptedException) x;
+                } else if (x instanceof RuntimeException) {
+                    throw (RuntimeException) x;
+                } else if (x != null) {
+                    throw new RuntimeException(x);
+                } else {
+                    throw new IllegalStateException();
+                }
+            }
+        }
+    }
+
+    /**
+     * Mask out query string or user info details in a URL.
+     * Useful in conjunction with {@code VirtualFile#toExternalURL}.
+     * @param url any URL
+     * @return the same, but with any {@link URL#getQuery} and/or {@link URL#getUserInfo} concealed
+     */
+    public static String sanitize(URL url) {
+        try {
+            URI orig = url.toURI();
+            return new URI(orig.getScheme(), orig.getUserInfo() != null ? "…" : null, orig.getHost(), orig.getPort(), orig.getPath(), orig.getQuery() != null ? "…" : null, orig.getFragment()).toString();
```

Comment:
I asked before, is `…` intended or should be `...` ? will unicode behave correctly?

Reply:
Either should be fine. This is just for use in build log messages, where we already often print `‘’` or `»` etc. (Cf. https://github.com/jenkinsci/workflow-job-plugin/pull/89.)